File: fullstack/go/price-calculator/managers/baseprice_test.go
package managers

import (
	"reflect"
	"testing"

	"redbubble.com/calculator/models"
)

func TestLookupSingleOptionItem(t *testing.T) {
	mgr := loadBasePriceManager(t)
	result, found := mgr.Lookup(
		"hoodie",
		map[string]string{
			"size":   "large",
			"colour": "white",
		},
	)
	if !found {
		t.Fatal()
	}

	expected := models.ProductBasePrice{
		ProductType: "hoodie",
		Options: map[string][]string{
			"colour": {"white"},
			"size":   {"large"},
		},
		Price: 3848,
	}
	if !reflect.DeepEqual(result, expected) {
		t.Fatal(result, "!=", expected)
	}
}

func TestLookupMultipleOptionItem(t *testing.T) {
	mgr := loadBasePriceManager(t)
	result, found := mgr.Lookup(
		"hoodie",
		map[string]string{
			"size":   "small",
			"colour": "dark",
		},
	)
	if !found {
		t.Fatal()
	}

	expected := models.ProductBasePrice{
		ProductType: "hoodie",
		Options: map[string][]string{
			"colour": {"white", "dark"},
			"size":   {"small", "medium"},
		},
		Price: 3800,
	}
	if !reflect.DeepEqual(result, expected) {
		t.Fatal(result, "!=", expected)
	}
}

func TestLookupZeroOptionItem(t *testing.T) {
	mgr := loadBasePriceManager(t)
	result, found := mgr.Lookup("leggings", nil)
	if !found {
		t.Fatal()
	}

	expected := models.ProductBasePrice{
		ProductType: "leggings",
		Options:     map[string][]string{},
		Price:       5000,
	}
	if !reflect.DeepEqual(result, expected) {
		t.Fatal(result, "!=", expected)
	}
}

File: fullstack/ruby/price_calculator/test/cart/models/base_price_test.rb
require 'minitest/autorun'
require 'test_helper'

require_relative '../../../lib/cart/models/product_base_price'

class BasePriceTest < Minitest::Test

  [nil, true, false, 'random string', '9', '-19'].each do |test_input|
    define_method("test_from_dict_invalid_#{test_input}") do
      assert_raises {
        ProductBasePrice.from_hash(test_input)
      }
    end
  end

  def test_from_valid_input
    test_input = {
      'product-type' => 'hoodie',
      'options' => {
        'colour' => ['white', 'dark'],
        'size' => ['small', 'medium']
      },
      'base-price' => 3800
    }

    result = ProductBasePrice.from_hash(test_input)
    assert_equal 'hoodie', result.product_type
    assert_equal Hash( 'colour' => ['white', 'dark'], 'size' => ['small', 'medium']), result.options
    assert 3800, result.price
  end
end
